drivers/pwm/pwm_nrfx.c

```diff
@@ -5,93 +5,258 @@
 
 struct pwm_nrfx_config {
 	nrfx_pwm_t pwm;
-	nrfx_pwm_config_t config;
+	nrfx_pwm_config_t initial_config;
 	nrf_pwm_sequence_t seq;
 };
 
 struct pwm_nrfx_data {
 	u16_t current[NRF_PWM_CHANNEL_COUNT];
-	u16_t top_value;
+	bool  pwm_instance_in_use;
+	u8_t  prescaler;
+	u16_t countertop;
+	u32_t period_cycles;
 };
 
+static u32_t pwm_period_check_and_set(const struct pwm_nrfx_config *pconfig,
+	struct pwm_nrfx_data *pdata, NRF_PWM_Type *pwm_instance, u32_t pwm,
+	u32_t period_cycles)
+{
+	if (pdata->pwm_instance_in_use) {
+		/* Succeed if requested period matches already used period */
+		if (period_cycles == pdata->period_cycles) {
+			return 0;
+		}
+
+		/* Fail if requested period != already running period */
+		LOG_ERR("Fail:requested period cycles:%d, != used %d\n",
+			period_cycles, pdata->period_cycles);
+		return -EINVAL;
+	}
+
+	/* Check if period_cycles is above COUNTERTOP MAX value, if so, we
+	 * have to see if we can change frequency to something that will fit
+	 */
+	if (period_cycles > PWM_COUNTERTOP_COUNTERTOP_Msk) {
+		/* See if there is a prescaler that will make it work: */
+		bool matching_prescaler_found = false;
+
+		/* Go through all available prescaler values on device.
+		 * nRF52832 has 0-7 (Div1 - Div128)
+		 */
+		for (u8_t prescaler = 0;
+				prescaler <= PWM_PRESCALER_PRESCALER_Msk;
+				prescaler++) {
+			u32_t new_countertop = period_cycles >> prescaler;
+
+			/* If we find value that fits, set it, continue */
+			if (new_countertop <= PWM_COUNTERTOP_COUNTERTOP_Msk) {
+				pdata->prescaler = prescaler;
+				pdata->countertop = new_countertop;
+				pdata->period_cycles = period_cycles;
+				matching_prescaler_found = true;
+				break;
+			}
+		}
+
+		/* Check if able to find matching prescaler and countertop */
+		if (matching_prescaler_found == false) {
+			LOG_ERR("Prescaler for period_cycles %d not found.\n",
+				period_cycles);
+			return -EINVAL;
+		}
+	} else {
+		/* If period_cycles fit with standard prescaler,
+		 * set it directly
+		 */
+		pdata->prescaler = PWM_PRESCALER_PRESCALER_DIV_1;
+		pdata->countertop = period_cycles;
+		pdata->period_cycles = period_cycles;
+	}
+
+	/* Write new PRESCALER and COUNTERTOP to PWM instance */
+	nrf_pwm_configure(pwm_instance,
+		pdata->prescaler,
+		pconfig->initial_config.count_mode,
+		pdata->countertop);
+
+	return 0;
+}
+
+static u8_t pwm_channel_map(const struct pwm_nrfx_config *pconfig, u32_t pwm)
+{
+	u8_t i;
+
+	/* Find pin, return channel number */
+	for (i = 0; i < NRF_PWM_CHANNEL_COUNT; i++) {
+		if (pconfig->initial_config.output_pins[i]
+			!= NRFX_PWM_PIN_NOT_USED
+		&&
+			(pwm == (pconfig->initial_config.output_pins[i]
+			& NRFX_PWM_PIN_MASK))) {
+			return i;
+		}
+	}
+
+	/* Return NRF_PWM_CHANNEL_COUNT to show that PWM pin was not found. */
+	return NRF_PWM_CHANNEL_COUNT;
+}
+
+static u16_t get_channel_pulse_cycle(u16_t current)
+{
+	return (current & NRFX_PWM_COMPARE_MASK);
+}
+
+static bool get_pin_inversion_state(u16_t config_output_pin)
```

Comment:
`u8_t config_output_pin`
Do we actually need a separate function for this? It is used only in one place and its result is assigned to a variable named `channel_inverted_state` so the masking could be performed directly there, and the code will still be readable (and even more clearly showing what is actually going on).

Reply:
Removed.